### PR TITLE
Add a DetailedList widget.

### PR DESCRIPTION
**Pull request by freakboy3742**

### CODE REVIEW

File: examples/detailedlist/detailedlist/app.py [hunk 1]
```diff
@@ -9,11 +11,13 @@ class ExampleDetailedListApp(toga.App):
     def on_select_handler(self, widget, row, **kwargs):
         self.label.text = 'You selected row: {}'.format(row) if row is not None else 'No row selected'
 
-    def on_refresh_handler(self, widget, **kwargs):
+    async def on_refresh_handler(self, widget, **kwargs):
+        self.label.text = 'Refreshing list...'
+        await asyncio.sleep(1)
```
Comment (phildini) [inline examples/detailedlist/detailedlist/app.py]: Lol wut. Fake refresh time?

File: src/cocoa/toga_cocoa/libs/__init__.py [hunk 2]
```diff
@@ -2,5 +2,4 @@
 from .core_graphics import *
 from .core_text import *
 from .foundation import *
-from .quartz import *
```
Comment (phildini) [inline src/cocoa/toga_cocoa/libs/__init__.py]: Is this an orphan from previous? Or am I missing where this was used?

Comment (phildini) [reply src/cocoa/toga_cocoa/libs/__init__.py]: Ah. I see now, deleted file below. (still not sure why it was deleted, perhaps that will become clear)

File: src/cocoa/toga_cocoa/widgets/detailedlist.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        # Create the column for the detailed list
+        column = NSTableColumn.alloc().initWithIdentifier('data')
+        self.detailedlist.addTableColumn(column)
+        self.columns = [column]
+
+        cell = TogaDetailedCell.alloc().init()
+        column.dataCell = cell
+
+        cell.editable = False
+        cell.selectable = False
```
Comment (phildini) [inline src/cocoa/toga_cocoa/widgets/detailedlist.py]: I think we want to be able to expose both of these as options to set `True` in the future, yes? Not saying we need to do so now, just thinking about future planning.

File: src/cocoa/toga_cocoa/widgets/detailedlist.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self.detailedlist.reloadData()
+
+    def change(self, item):
+        self.detailedlist.reloadData()
+
+    def remove(self, item):
+        self.detailedlist.reloadData()
+
+    def clear(self):
+        self.detailedlist.reloadData()
```
Comment (phildini) [inline src/cocoa/toga_cocoa/widgets/detailedlist.py]: It "smells weird" that all of these methods are calling the same underlying method. Do we expect these to be overridden in use?

File: src/cocoa/toga_cocoa/widgets/internal/cells.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+from rubicon.objc import *
+
+from ...libs import *
```
Comment (phildini) [inline src/cocoa/toga_cocoa/widgets/internal/cells.py]: I have literally never seen the `...` import syntax, and couldn't find a google term to prove what it did. I'm assuming it imports from... two directories up? And I'd assume this code runs, so I see that it works, I'm just filled with a momentary "wat".

**Comment by phildini** [comment]
I think is probably close to read-to-ship, and I'd be fine giving my approval now if you want it and you're ready for it.

There's a couple smells I highlighted, especially around the interface we're exposing for the widgets and what the expected default behavior is. For me, this could be fixed with an advanced example or an advanced "how to use" doc.

File: examples/detailedlist/detailedlist/app.py (hunk 1 above)
Comment (freakboy3742) [reply examples/detailedlist/detailedlist/app.py]: Yes - it's a demo app; without a background server to get actual data, we need *something* to fake the load time. I should add a comment to explain though.

File: src/cocoa/toga_cocoa/libs/__init__.py (hunk 2 above)
Comment (freakboy3742) [reply src/cocoa/toga_cocoa/libs/__init__.py]: The Quartz module was a copy-pasta artefact from the code I used to bootstrap Rubicon. Turns out we weren't using any of it.

File: src/cocoa/toga_cocoa/widgets/detailedlist.py (hunk 3 above)
Comment (freakboy3742) [reply src/cocoa/toga_cocoa/widgets/detailedlist.py]: Definitely for selectable; possibly for editable. I've set them explicitly so future code explorers know where to start looking.

Comment (freakboy3742) [reply src/cocoa/toga_cocoa/widgets/detailedlist.py]: (although, interestingly... selectable doesn't seem to be preventing selection...)

File: src/cocoa/toga_cocoa/widgets/internal/cells.py (hunk 5 above)
Comment (freakboy3742) [reply src/cocoa/toga_cocoa/widgets/internal/cells.py]: Every `.` is "one level higher up". `.` is "current module", `..` is parent module, `...` is grandparent module, and so on. However, I suppose you're right - there's no reason to *not* use `toga_cocoa.libs` here.

File: src/cocoa/toga_cocoa/widgets/detailedlist.py (hunk 4 above)
Comment (freakboy3742) [reply src/cocoa/toga_cocoa/widgets/detailedlist.py]: No - it's an artefact of providing a single interface that provides enough detail for all backends to use.

Cocoa has a very purist "delegated data source" model for table/tree rendering. You don't add and remove elements to the visual tree/table; you specify a delegate that answers the question "what should I display at row X, column Y" on demand. `reloadData` is the way you tell the table "BTW - something visible has changed, please do a redraw". No matter the change, the visual response is "... and redraw".

GTK and Winforms, however, require a specific response to "add", "remove", "change", and so on, that insert, remove and modify specific nodes in the tree. So the interface needs to provide all the methods to support GTK and and Winforms; Cocoa (and FTR, iOS) just need to say "yup, got that; rerender".

**Pull request merged**